proxy/src/control/discovery.rs

```diff
@@ -348,52 +348,17 @@ where
 
     fn poll_destinations(&mut self) {
         for (auth, set) in &mut self.destinations {
-            let needs_reconnect = 'set: loop {
-                let poll_result = match set.query {
-                    None |
-                    Some(DestinationServiceQuery::NeedsReconnect) => {
-                        continue;
-                    },
-                    Some(DestinationServiceQuery::ConnectedOrConnecting{ ref mut rx }) => {
-                        rx.poll()
-                    }
-                };
-
-                match poll_result {
-                    Ok(Async::Ready(Some(update))) => match update.update {
-                        Some(PbUpdate2::Add(a_set)) =>
-                            set.add(
-                                auth,
-                                a_set.addrs.iter().filter_map(
-                                    |addr| addr.addr.clone().and_then(pb_to_sock_addr))),
-                        Some(PbUpdate2::Remove(r_set)) =>
-                            set.remove(
-                                auth,
-                                r_set.addrs.iter().filter_map(|addr| pb_to_sock_addr(addr.clone()))),
-                        Some(PbUpdate2::NoEndpoints(no_endpoints)) =>
-                            set.no_endpoints(auth, no_endpoints.exists),
-                        None => (),
-                    },
-                    Ok(Async::Ready(None)) => {
-                        trace!(
-                            "Destination.Get stream ended for {:?}, must reconnect",
-                            auth
-                        );
-                        break 'set true;
-                    }
-                    Ok(Async::NotReady) => break 'set false,
-                    Err(err) => {
-                        warn!("Destination.Get stream errored for {:?}: {:?}", auth, err);
-                        break 'set true;
-                    }
+            let query = set.query.take();
+            set.query = if let Some(DestinationServiceQuery::ConnectedOrConnecting{ rx }) = query {
+                let new_query = set.poll_destination(auth, rx);
+                if let DestinationServiceQuery::NeedsReconnect = new_query {
+                    set.reset_on_next_modification();
+                    self.reconnects.push_back(auth.clone());
                 }
-
+                Some(new_query)
+            } else {
+                query
```

Comment:
tioli: this would read slightly more naturally to me as:

```rust
            set.query = match set.query.take() {
                Some(DestinationServiceQuery::ConnectedOrConnecting{ rx }) => {
                    let new_query = set.poll_destination(auth, rx);
                    if let DestinationServiceQuery::NeedsReconnect = new_query {
                        set.reset_on_next_modification();
                        self.reconnects.push_back(auth.clone());
                    }
                    Some(new_query)
                }
                query => query
             };
```

Reply:
> tioli: this would read slightly more naturally to me as:

Thanks. I reformulated it the way you suggested.